src/cryptonote_core/service_node_list.cpp

```diff
@@ -250,11 +249,13 @@ namespace service_nodes
     if (total > STAKING_SHARES)
       return false;
 
-    if (is_service_node(service_node_key))
+    crypto::hash hash;
+    if (!get_registration_hash(service_node_addresses, service_node_shares, expiration_timestamp, hash))
+      return false;
+    if (!crypto::check_signature(hash, service_node_key, signature))
+      return false;
+    if (expiration_timestamp < block_timestamp)
       return false;
-
-    // TODO: check service_node_key has signed service node addresses and
-    // service node shares and timestamp, and that timestamp is not old.
 
     cryptonote::keypair gov_key = cryptonote::get_deterministic_keypair_from_height(1);
 
```

Comment:
I noticed below in the loop `get_reg_tx_staking_output_contribution(tx, **<index>**, derivation, hwdev)` was fixed previously so that you don't address out of bounds, but

`if (tx.vout.size() < service_node_addresses.size())`

Implies tx.vout.size() is >= service_node_addresses.size(), so shouldn't you need to check all the outputs against each address derivation, not just (0->number of service node addresses) which isn't necessarily the number of output in the TX.

Reply:
We could do it that way, but I chose not to. So a registration to address1 shares1 address2 shares2 etc. Must be constructed with address1 as the first output, shares2 as the second, and so on.

Reply:
Okay, I guess its their fault if they lock up their funds trying anything that's not the way we're checking it.